streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java

```diff
@@ -0,0 +1,461 @@
+                maxWaitTime = System.currentTimeMillis() + 30000;
+                addRecord(record, recordPerTopicPerPartition);
+            }
+
+            if (receivedRecords.count() > 0) {
+                allRecordsReceived =
+                    receivedAllRecords(
+                        recordPerTopicPerPartition.get("data"),
+                        recordPerTopicPerPartition.get("echo"),
+                        committedOffsets);
+            }
+        }
+
+        if (!allRecordsReceived) {
+            throw new RuntimeException("FAIL: did not receive all records after 30 sec idle time.");
+        }
+
+        return recordPerTopicPerPartition;
+    }
+
+    private static void addRecord(final ConsumerRecord<byte[], byte[]> record,
+                                  final Map<String, Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>>> recordPerTopicPerPartition) {
+
+        final String topic = record.topic();
+        final TopicPartition partition = new TopicPartition(topic, record.partition());
+
+        if ("data".equals(topic)
+            || "echo".equals(topic)
+            || "min".equals(topic)
+            || "sum".equals(topic)) {
+
+            Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> topicRecordsPerPartition
+                = recordPerTopicPerPartition.get(topic);
+
+            if (topicRecordsPerPartition == null) {
+                topicRecordsPerPartition = new HashMap<>();
+                recordPerTopicPerPartition.put(topic, topicRecordsPerPartition);
+            }
+
+            List<ConsumerRecord<byte[], byte[]>> records = topicRecordsPerPartition.get(partition);
+            if (records == null) {
+                records = new ArrayList<>();
+                topicRecordsPerPartition.put(partition, records);
+            }
+            records.add(record);
+        } else {
+            throw new RuntimeException("FAIL: received data from unexpected topic: " + record);
+        }
+    }
+
+    private static boolean receivedAllRecords(final Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> supersetExpectedRecords,
+                                              final Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> receivedRecords,
+                                              final Map<TopicPartition, Long> committedOffsets) {
+        if (supersetExpectedRecords == null
+            || receivedRecords == null
+            || supersetExpectedRecords.keySet().size() < committedOffsets.keySet().size()
+            || receivedRecords.keySet().size() < committedOffsets.keySet().size()) {
+
+            return false;
+        }
+
+        for (final Map.Entry<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords : receivedRecords.entrySet()) {
+            final Long committed = committedOffsets.get(partitionRecords.getKey());
+            if (partitionRecords.getValue().size() < (committed == null ? 0 : committed)) {
```

Comment:
Is this okay? If committed == null it means no offsets get committed for this partition in `getCommittedOffsets`, and we should check that we do not get any data in `partitionRecords ` but currently this check will always pass even if `partitionRecords.getValue().size() > 0`.